lang/funcs/core/os/exec_func.go

```diff
@@ -0,0 +1,164 @@
+// Mgmt
+// Copyright (C) 2013-2022+ James Shubin and the project contributors
+// Written by James Shubin <james@shubin.ca> and the project contributors
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package coreos
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"os/exec"
+	"sync"
+
+	"github.com/purpleidea/mgmt/lang/funcs"
+	"github.com/purpleidea/mgmt/lang/interfaces"
+	"github.com/purpleidea/mgmt/lang/types"
+)
+
+func init() {
+	funcs.ModuleRegister(ModuleName, "exec", func() interfaces.Func { return &ExecFunc{} })
+}
+
+// ExecFunc runs a string as a shell command, then produces each line from
+// stdout. If the input string changes, then the commands are executed one after
+// the other and the concatenation of their outputs is produced line by line.
+//
+// Note that in the likely case in which the process emits several lines one
+// after the other, the downstream resources might not run for every line unless
+// the "Meta:realize" metaparam is set to true.
+type ExecFunc struct {
+	init *interfaces.Init
+
+	// Used to shutdown the node, killing the current process if needed.
+	// We only run one process at a time.
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// ArgGen returns the Nth arg name for this function.
+func (obj *ExecFunc) ArgGen(index int) (string, error) {
+	seq := []string{"shell_command"}
+	if l := len(seq); index >= l {
+		return "", fmt.Errorf("index %d exceeds arg length of %d", index, l)
+	}
+	return seq[index], nil
+}
+
+// Validate makes sure we've built our struct properly. It is usually unused for
+// normal functions that users can use directly.
+func (obj *ExecFunc) Validate() error {
+	return nil
+}
+
+// Info returns some static info about itself.
+func (obj *ExecFunc) Info() *interfaces.Info {
+	return &interfaces.Info{
+		Pure: false, // definitely false
+		Memo: false,
+		Sig:  types.NewType("func(shell_command str) str"),
+		Err:  obj.Validate(),
+	}
+}
+
+// Init runs some startup code for this function.
+func (obj *ExecFunc) Init(init *interfaces.Init) error {
+	obj.init = init
+	obj.ctx, obj.cancel = context.WithCancel(context.Background())
+	return nil
+}
+
+// Stream returns the changing values that this func has over time.
+func (obj *ExecFunc) Stream() error {
+	var wg sync.WaitGroup
```

Comment:
Sneaky, but you should actually do:

`wg := &sync.WaitGroup{}`

If you look at the docs, I think it suggests this way is safer so that we can copy this thing around. But perhaps I've misunderstood things. All I know is using the aforementioned form seems to have always worked better.